# Task 3: Data Cleaning

## Project Overview
This project focuses on cleaning and preparing the Online Retail dataset for further analysis. The dataset is inspected for missing values, duplicate records, inconsistent or invalid values, and other data quality issues.

In [28]:
import pandas as pd

In [29]:
df = pd.read_excel('data/Online Retail.xlsx')

In [30]:
# Save original data quality metrics
original_rows = len(df)
original_missing= df.isnull().sum().sum()
original_duplicates= df.duplicated().sum()

print("Before Cleaning")
print("Total Rows:",original_rows)
print("Total Missing values:",original_missing)
print("Duplicated Rows:",original_duplicates)

Before Cleaning
Total Rows: 541909
Total Missing values: 136534
Duplicated Rows: 5268


##  Data Inspection


In [31]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [32]:
df.shape

(541909, 8)

In [33]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


In [34]:
df.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID            float64
Country                   str
dtype: object

## Data Type validation
Data types were checked and found suitable for the dataset. No unnecessary conversions were made.

## Handling Missing Values

In [35]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [36]:
missing_percentage = (df.isnull().sum() / len(df)) * 100
missing_percentage

InvoiceNo       0.000000
StockCode       0.000000
Description     0.268311
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     24.926694
Country         0.000000
dtype: float64

In [37]:
# Missing Description values were replaced with Unknown, while rows with missing CustomerID were removed because the customer information was unavailable.

df['Description'] = df['Description'].fillna('Unknown')

In [38]:
df= df.dropna(subset=['CustomerID'])

In [39]:
df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

## Duplicate remove

In [40]:
df.duplicated().sum()

np.int64(5225)

In [41]:
# Duplicate rows were identified and removed to prevent from affecting the analysis

df = df.drop_duplicates()

## Checking Invalid Values
Quantity and UnitPrice are examined for unusual values. Non-positive unit prices are removed. Negative quantities are also checked because they may represent product returns or cancelled transactions.

In [42]:
df[['Quantity','UnitPrice']].describe()

,Quantity,UnitPrice
count,401604.000000,401604.000000
mean,12.183273,3.474064
std,250.283037,69.764035
min,-80995.000000,0.000000
25%,2.000000,1.250000
50%,5.000000,1.950000
75%,12.000000,3.750000
max,80995.000000,38970.000000


In [43]:
print('Negative Quantity:',(df['Quantity']<0).sum())
print('Zero UnitPrice:',(df['UnitPrice']==0).sum())

Negative Quantity: 8872
Zero UnitPrice: 40


In [44]:
df[df['UnitPrice'] == 0][['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'UnitPrice']].head(10)

,InvoiceNo,StockCode,Description,Quantity,UnitPrice
9302,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,0.0
33576,539263,22580,ADVENT CALENDAR GINGHAM SACK,4,0.0
40089,539722,22423,REGENCY CAKESTAND 3 TIER,10,0.0
47068,540372,22090,PAPER BUNTING RETROSPOT,24,0.0
47070,540372,22553,PLASTERS IN TIN SKULLS,24,0.0
56674,541109,22168,ORGANISER WOOD ANTIQUE WHITE,1,0.0
86789,543599,84535B,FAIRY CAKES NOTEBOOK A6 SIZE,16,0.0
130188,547417,22062,CERAMIC BOWL WITH LOVE HEART DESIGN,36,0.0
139453,548318,22055,MINI CAKE STAND HANGING STRAWBERY,5,0.0
145208,548871,22162,HEART GARLAND RUSTIC PADDED,2,0.0


In [45]:
df = df[df['UnitPrice'] > 0]
(df['UnitPrice'] <= 0).sum()

np.int64(0)

In [46]:
df[df['Quantity'] < 0][['InvoiceNo', 'StockCode', 'Description', 'Quantity']].head(10)

,InvoiceNo,StockCode,Description,Quantity
141,C536379,D,Discount,-1
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12
241,C536391,22553,PLASTERS IN TIN SKULLS,-24
939,C536506,22960,JAM MAKING SET WITH JARS,-6


In [47]:
(df.loc[df['Quantity'] < 0, 'InvoiceNo'].astype(str).str.startswith('C')).sum()

np.int64(8872)

In [48]:
df = df[df['Quantity'] > 0]
(df['Quantity'] < 0).sum()

np.int64(0)

In [49]:
print('Shape:',df.shape),
print('Missing Value',df.isnull().sum()),
print('Duplicates:',df.duplicated().sum()),
print('Negative Quantity:',(df['Quantity']<0).sum()),
print('Non-Positive UnitPrice:',(df['UnitPrice']<=0).sum())

Shape: (392692, 8)
Missing Value InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64
Duplicates: 0
Negative Quantity: 0
Non-Positive UnitPrice: 0


## Outlier Detection Using IQR Method

In [50]:
for col in ['Quantity','UnitPrice']:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)

    IQR = Q3-Q1
    lower = Q1-1.5*IQR
    upper = Q3+1.5*IQR

    outliers= df[(df[col]<lower)| (df[col]>upper)]

    print(f'\n{col}')
    print('Lower limit:',lower)
    print('Upper limit:',upper)
    print('Number of outliers:',len(outliers))


Quantity
Lower limit: -13.0
Upper limit: 27.0
Number of outliers: 25616

UnitPrice
Lower limit: -2.5
Upper limit: 7.5
Number of outliers: 34112


### Outlier Detection and Handling

The IQR method identified 25,616 potential outliers in Quantity and 34,112 in UnitPrice. These values were identified as potential outliers and were not directly removed using the IQR method. Separately, rows with non-positive UnitPrice and non-positive Quantity were removed during data cleaning to exclude these records from the final dataset. This may exclude some genuine product returns or cancellations represented by negative quantities.

### Before-and-After Data Quality Summary

Compared the dataset before and after cleaning to evaluate changes in row count, missing values, and duplicate records.


In [51]:
print("After Cleaning")
print("Total Rows:", len(df))
print("Total Missing Values:", df.isnull().sum().sum())
print("Duplicate Rows:", df.duplicated().sum())

After Cleaning
Total Rows: 392692
Total Missing Values: 0
Duplicate Rows: 0


In [54]:
summary = pd.DataFrame({
    'Metric': ['Total Rows', 'Missing Values', 'Duplicate Rows','Data Types'],
    'Before Cleaning': [541909, 136534, 5268,'Inspected'],
    'After Cleaning': [len(df),df.isnull().sum().sum(),
        df.duplicated().sum(),'Validated']})
summary

,Metric,Before Cleaning,After Cleaning
0,Total Rows,541909,392692
1,Missing Values,136534,0
2,Duplicate Rows,5268,0
3,Data Types,Inspected,Validated


In [53]:
df.to_csv('data/Online_Retail_Cleaned.csv', index=False)